torch/optim/zeroptimizer.py

```diff
@@ -0,0 +1,524 @@
+    else:
+        global_rank = dist.distributed_c10d._get_global_rank(group, rank)  # type: ignore
+    return global_rank
+
+
+class ZeROptimizer(Optimizer):
+    """Wraps an arbitrary :class:`optim.Optimizer <torch.optim.Optimizer>`
+    optimizer and shards its state as described by ZeRO_.
+    ::
+
+        opt = ShardedOptimizer(params, optim=torch.optim.Adam, lr=0.01)
+
+    .. _ZeRO: https://arxiv.org/abs/1910.02054
+
+    We use a greedy algorithm to pack a number of parameters
+    at each rank. Each parameter belongs to a single rank and
+    is not divided among rank.
+
+    After each rank completed their parameter update, they broadcast
+    the new version of the parameters to all other ranks to synchronize
+    the parameters for next round forward/backward computation.
+
+    Args:
+        params (list of tensors):
+            parameters to be optimized
+    Keyword Args:
+        optim (torch.nn.Optimizer):
+            optimizer to shard
+        group (group):
+            torch.distributed group (default: group.WORLD)
+        broadcast_buffer_size (int):
+            the size of the buffer used to batch the small parameter tensors (default 128k).
+    """
+
+    #: The optimizer used for a given shard
+    optim: Optimizer
+
+    in_super_constructor: bool
+
+    def __init__(
+        self,
+        params: _params_t,
+        optim: Type[Optimizer],
+        group: Optional[Any] = None,
+        broadcast_buffer_size: int = 2 ** 17,
+        **default: Any,
+    ):
+        # Hold all the model params in the root .param_groups
+        self.in_super_constructor = True
+        super().__init__(params, default)
+        self.in_super_constructor = False
+
+        # Partition information. lazy evaluation, computed if requested
+        self._per_device_params: OrderedDict[
+            torch.device, List[List[Parameter]]
+        ] = OrderedDict()  # device, rank, params
+        self._param_rank: Dict[torch.Tensor, int] = {}
+        self._partition_parameters: List[List[dict]] = []
+
+        # Build the wrapped optimizer, responsible for a shard of the params
+        self.group = group if group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.group)
+        self.rank = dist.get_rank(self.group)
+        self.global_rank = _get_global_rank(self.group, self.rank)
+
+        self.optim = optim(self.partition_parameters()[self.rank], **default)
+
+        # - Sync local and global param_groups keys
+        for global_group, local_group in zip(self.param_groups, self.optim.param_groups):
+            for k, v in local_group.items():
+                if k != "params":
+                    global_group[k] = v
+
+        #  Optional consolidated optimizer state
+        self._all_states: List[Dict[str, Any]] = []
+
+        # Current default device is set by the parameters allocated to this rank
+        self._device = self.partition_parameters()[self.rank][0]["params"][0].device
```

Comment:
What's the second dimension of the `self.partition_parameters()`? I mean the one after `self.rank`. Is it device? If yes, we can avoid that, as new features in the distributed package can assume that each rank exclusively works on one device. And we are going to retire the single-process multi-device mode in DDP as well.

Reply:
a sync with @msbaines would be nice, a key assumption in this code is that ranks can have parameters spread across multiple devices, cannot be removed lightly

Reply:
I am a little surprised that the legacy single-process multi-device mode is indeed used here. @msbaines could you please elaborate on why do you need this mode?

Reply:
I just realized that the above only replied to part of your question. The dimensions look like `self.partition_parameters[rank][param_group]`, and each param could be on a different device, which is a hard requirement to keep the sharded optimizer compatible with some techniques.
This self._device is kind of a bandaid though, for some methods we need to have a handle to 'a' device (for instance for state broadcasting), so this saves a handle to a device which works

Reply:
> I am a little surprised that the legacy single-process multi-device mode is indeed used here. @msbaines could you please elaborate on why do you need this mode?

I can reply to that, combining OSS + model parallel for instance means that OSS ranks could own several devices, that's a use case actively being used in fairseq